src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -625,102 +610,73 @@ namespace xclhwemhal2 {
     if (xclemulation::config::getInstance()->isSharedFmodel() && !mVersalPlatform) {
       setenv("SDX_USE_SHARED_MEMORY","true",true);
     }
-    // iterate cores
-    count = 0;
-    for (auto& xml_core : xml_project.get_child("project.platform.device"))
-    {
-      if (xml_core.first != "core")
-        continue;
-      if (++count > 1)
-      {
-        //Give error and return from here
-      }
-    }
 
-    std::vector<std::string> kernels;
-
-    // iterate kernels
-    for (auto& xml_kernel : xml_project.get_child("project.platform.device.core"))
+    auto xclbin_object = new xrt::xclbin(args.m_top);
+    auto kernels = xclbin_object->get_kernels();
+    std::map<uint64_t, KernelArg> kernelArgInfo;
+    std::string instance_name = "";
+    uint64_t base_address = 0;
+    for (auto kernel : kernels)
     {
-      if (xml_kernel.first != "kernel")
-        continue;
-
-      std::string kernelName = xml_kernel.second.get<std::string>("<xmlattr>.name");
-      kernels.push_back(kernelName);
-      int address_range = 0;
-      std::string instanceName;
-
-      if (mLogStream.is_open())
-         mLogStream << __func__ << " Filling kernel " << kernelName << " info from xclbin.xml" << std::endl;
-
-      for (auto& xml_kernel_info : xml_kernel.second)
+      // get properties of each kernel object
+      auto props = xrt_core::xclbin_int::get_properties(kernel);
+      //get CU's of each kernel object
+      auto cus = kernel.get_cus();
+      //iterate over CU's to get arguments
+      for (auto cu :cus)
       {
-        std::map<uint64_t, KernelArg> kernelArgInfo;
-        if (xml_kernel_info.first == "arg")
+        auto args = cu.get_args();
```

Comment:
Move variables to where they are used.  `args` is not used until line 644 and can even be combined with the for-loop there.